### ISSUE TITLE
7z is not considered a word in PowerShell

### ISSUE THREAD

**Issue by GabrielFrigo4** [ScintillaOrg/lexilla#118]
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12465

In notepad++ when I write the word 7z it is not highlighted, I reported this to the developers of notepad++ and they told me to report here.

![image](https://user-images.githubusercontent.com/66432268/201713673-f1528565-2b86-49d5-bff2-103ab8a790dc.png)

**Comment by GabrielFrigo4** [comment]
> The PowerShell lexer is seeing a digit and treating it as a numeric literal (number). https://learn.microsoft.com/en-us/powershell/scripting/lang-spec/chapter-02?view=powershell-7.2#23511-integer-literals

Does that make sense, but wasn't it just to happen with certain digits, with for example 7d? 7z is conciderado a word by powershel 7.3, and 7d a number

![Captura de ecrã de 2022-11-14 18-19-18](https://user-images.githubusercontent.com/66432268/201767666-102b42e6-26fc-4802-a17a-1f0af9b4621f.png)

**Comment by nyamatongwe** [comment]
Unsure how to lex this like PowerShell - it may be performing a search for commands in the path and prioritizing them. There may also be different behaviour at line start when it is expecting a command than in other locations.

Since I don't personally use PowerShell, any code changes will have to be contributed by others.

**Comment by mpheath** [comment]
Styling in Terminal:

```
7a yellow
7b yellow
7c yellow
7d white
7e yellow
7f yellow
7x yellow
7y yellow
7z yellow
```

`7d` is white. I thought `7[a-f]` be white as being hexidecimal and the remaining `7[g-z]` be yellow. The styling conditions are unknown due to this. Tested with Windows 11 in VirtualBox.

Terminal just adds more confusion. It does not colour numbers, example:

```
$i = 7
```

`7` is white. White is the default colour. This is not like the Lexilla powershell lexer which identifies numbers and styles them.

Notepad++ has a plugin known as [EnhanceAnyLexer](https://github.com/Ekopalypse/EnhanceAnyLexer)

```ini
[powershell]
0x99DFCE = ^\h*\b(?:gcc|7z|as|nasm)\b
```

The Lexilla lexer recognizes numbers and strings before identifiers such as cmdlets, functions, aliases... so a leading number is going to be always recognized as a number. EnhanceAnyLexer can style the `7z` to the defined color in it's configuration. This is a workaround concept though changing the lexer to style `7z` based on what condition is currently unknown to me. Someone with better powershell knowledge may know how to improve the lexer.

**Comment by nyamatongwe** [comment]
For `7d`, `d` is a `decimal-type-suffix` from the definition of real literals.
https://learn.microsoft.com/en-us/powershell/scripting/lang-spec/chapter-02?view=powershell-7.2#23512-real-literals

**Comment by mpheath** [comment]
@nyamatongwe Thanks for the link. This link also has good numeric literal information.

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_numeric_literals?view=powershell-7.3#examples

This is a regex possible solution. I have not seen regex used in the lexers so whether it is allowed is unknown so decided to test if was possible to do lexing numeric literals with regex.

The regex tries to match the string from `sc.GetCurrentLowered()` and `MakeLowerCase(sc.ch)`. The former appears to not be so current for an unknown reason.

Due to character by character (C&C) processing by the lexer, regex matching success until the end of the (numeric) literal could lead to invalid literals being styled as literals. Example `0x` matches as start of a hexidecimal, though if followed by space, remains styled as a literal. Another function call could invalidate the `0x` though needs more regex patterns which complicates the lexer code even more. Another example, `mb` suffix with the `m` missing could be a match. The ECMAScript regex does not support look behinds so cannot validate with the same pattern, although group matches might be possible to assist with suffix validation. C&C causes me to create wierd unreliable patterns to maintain a positive match.

Removed the leading `.` recognition of a number as it complicates the regex patterns. Leading `-` and leading `+` are also excluded for the same reason.

Accounting for suffixes causes lexer maintenance for newer Powershell releases though without, the literals look rather bad as the gif shows.

<details>

<summary>Click here to view the gif.</summary>

Before and after as displayed in SciTE.

![powershell0](https://user-images.githubusercontent.com/58158242/203720479-d1767880-b7b9-4af8-8f93-cf1d8fde86c3.gif)

The "after" is more red characters than black characters. Note `7z` fails regex match so changes to black.

<hr>

</details>

If regex is not allowed in lexers, then any further development with this regex code will end.

[ps_numeric_literals.zip](https://github.com/ScintillaOrg/lexilla/files/10081750/ps_numeric_literals.zip)

**Comment by nyamatongwe** [comment]
Regular expressions can be used if you want although its the first time in Lexilla so there may be issues. They may also be slower than low level code.

The use of `sprint` in this patch is strange as it is just copying a char into a char array. The 4 makes it appear that something longer than 2 char could occur here. Something like this would be more obvious although it needs a cast:
~~~c++
char c[2]{};
c[0] = static_cast<char>(MakeLowerCase(sc.ch));
// or even
const char c[2] { static_cast<char>(MakeLowerCase(sc.ch)) };
~~~

Since `IsNumericLiteral` returns `bool`, the `bool` literals `true` and `false` should be used for returns instead of `0` and `1`. `regex_match` returns a `bool` so the final 4 lines can be shortened to
~~~c++
return std::regex_match(s, literal_regex);
~~~

**Comment by mpheath** [comment]
@nyamatongwe Thanks for the cast code as `sprintf` achieved what previous attempts failed with an int to lowercase string. Sometimes a bit of a struggle for me with little c++ experience. Your c++ knowledge certainly helps me.

A more passive concept of numeric literal is start with digit and within a certain group of characters continue with minimal code and perhaps only use possibly regex at the end for 100% validation. This may make it more efficient.

I may look into and learn from Notepad2 code as it is quite passive though too passive as `7z` styles as numeric literal which this issue report is about. So, a bit of this and a bit of that might make a better alternative. Will reconsider the options and try again.

**Comment by mpheath** [comment]
@nyamatongwe

This alternative does not use regex. It is low level code that does similar to the previous regex code so no numeric literal validation. A separate syntax checker for Powershell programmers is advised.

Thanks to Notepad2, some ideas helped with improvements with range of numbers not colouring the dots, exponents (e) might be signed -/+... . Notepad2 includes the whole alphabet while this code only allows only valid letters for hex, suffix... so less chance of matching words like `7z`.

<details>

<summary>Click to view the gif:</summary>

Before and after as displayed in SciTE.

![powershell1](https://user-images.githubusercontent.com/58158242/203964381-6ffe8a46-9be1-4beb-9412-828604abb189.gif)

The "after" is more red characters than black characters. Note `7z` fails regex match so changes to black.

<hr>

</details>

[ps_numeric_literals2.zip](https://github.com/ScintillaOrg/lexilla/files/10091718/ps_numeric_literals2.zip)

**Comment by nyamatongwe** [comment]
Committed the non-regex change.

**Comment by mpheath** [comment]
@nyamatongwe

This line on failure or numeric literal identification

https://github.com/ScintillaOrg/lexilla/blob/621fefc5333752ce0fb2ea0d3df62c1d09b4c836/lexers/LexPowerShell.cxx#L143

causes this in the test

https://github.com/ScintillaOrg/lexilla/blob/621fefc5333752ce0fb2ea0d3df62c1d09b4c836/test/examples/powershell/NumericLiterals.ps1.styled#L33-L35

Today I reconsidered the change of SCE_POWERSHELL_DEFAULT, to instead to SCE_POWERSHELL_IDENTIFIER. This might give the `7z` another chance of being identified from a keyword list.

So I tested with the focus of this part of the test file 

```
# 7d is numeric while 7z is default+identifier
7d
7z
7z|7d
7z | 7d
```

which is a little more text, though lets me know if the pipe or space is processed correct. This is the result.

```
{1}# 7d is numeric while 7z is default+identifier{0}
{4}7d{0}
{7}7z{0}
{7}7z{6}|{4}7d{0}
{7}7z{0} {6}|{0} {4}7d{0}
```

`{7}`=identifier, `{6}`=operator, `{4}`=number

This looks like a better result than the default+identifier with `7z`. This change may also allow `7z` to be recognized as a user keyword or another defined keyword.

In SciTE.properties, I changed the user-defined keyword property

https://github.com/ScintillaOrg/lexilla/blob/621fefc5333752ce0fb2ea0d3df62c1d09b4c836/test/examples/powershell/SciTE.properties#L8

to

```
keywords5.*.ps1=lexilla 7z
```

and the test result is

```
{1}# 7d is numeric while 7z is default+identifier{0}
{4}7d{0}
{12}7z{0}
{12}7z{6}|{4}7d{0}
{12}7z{0} {6}|{0} {4}7d{0}
```

Updating the SciTE.properties is not a goal. Just the identification is enough to prove it can become a keyword.

Perhaps the style should change to SCE_POWERSHELL_IDENTIFIER due to what I consider as better results.

The fix is the last commit so should be able to amend the commit, so if seems good, then with hopefully same formatting... being correct then the files to amend:

[ps_numeric_literals3.zip](https://github.com/ScintillaOrg/lexilla/files/10109500/ps_numeric_literals3.zip)

**Comment by nyamatongwe** [comment]
Seems fine to me but I'm not a PowerShell user. There could be a test for highlighting "7z" or similar as a keyword.

Since the earlier commit may have been cloned by others, its simpler and safer to add another commit.

**Comment by mpheath** [comment]
@nyamatongwe 

I did several weeks of learning Powershell until I abandoned it due to its issues several years ago. I'm reluctant to use it again while other options exist. Enough said without going into opinions. 

Yeah, I tend to agree as force pushing to sync an amended commit with an online repo is disturbing as bad consequences could happen.

So if you want a test with `7z` as a user defined keyword

[ps_numeric_literals4.zip](https://github.com/ScintillaOrg/lexilla/files/10119666/ps_numeric_literals4.zip)

Earlier testing left me confused as how mixed up styles can be with the like `default`+`identifier` style. Consider search in styles option with **Find** for the word `7z` with(out) mixed styles. More obvious is the styling of `7z` that the OP may want to highlight.

Off-topic: Some property values for the `identifier` style in various languages is not set a value in SciTE which could possibly be set to style 0 as default. This caused confusion in me to also see.

So it is taking a little time to reach clarity about the `identifier` style and where it should be applied. In this case, change to `identifier` style seems to be a benefit just because of one word `7z`.
